### PR TITLE
CASSANDRA-19634: Improve test coverage for downed instances

### PR DESCRIPTION
**Pull request by frankgh**
Patch by Francisco Guerrero; Reviewed by Yifan Cai for CASSANDRA-19634

### CODE REVIEW

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java [hunk 1]
```diff
@@ -428,9 +436,11 @@ protected Object[][] queryAllData(QualifiedName table)
      * @param consistencyLevel the consistency level to use for querying the data
      * @return all the data queried from the table
      */
-    protected Object[][] queryAllData(QualifiedName table, ConsistencyLevel consistencyLevel)
+    protected Object[][] queryAllData(QualifiedName table, String consistencyLevel)
```
Comment (yifan-c) [inline cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java]: nit: prefer type over string

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java [hunk 2]
```diff
@@ -486,20 +496,23 @@ public static Cluster createDriverCluster(ICluster<? extends IInstance> dtest)
     static class IntegrationTestModule extends AbstractModule
     {
         private static final Logger LOGGER = LoggerFactory.getLogger(IntegrationTestModule.class);
-        private final ICluster<? extends IInstance> cluster;
+        private final Iterable<? extends IInstance> instances;
```
Comment (yifan-c) [inline cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java:498]: `ICluster` is already the `Iterable`. Why changing to the new type here? 
Beside, `ICluster` has convenient methods to return a stream of the instances.

File: cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteDownInstanceTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import static org.apache.cassandra.testing.TestUtils.TEST_KEYSPACE;
+
+/**
+ * Tests bulk writes in different scenarios when Cassandra instances are down
+ */
+class BulkWriteDownInstanceTest extends SharedClusterSparkIntegrationTestBase
+{
+    Set<IInstance> downInstances = new HashSet<>();
+
+    @ParameterizedTest(name = "{index} => instanceDownCount={0} {1}")
```
Comment (yifan-c) [inline cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteDownInstanceTest.java]: nit: so it outputs `instanceDownCount=0, readCL=ONE, writeCL=ALL`
```suggestion
    @ParameterizedTest(name = "{index} => instanceDownCount={0}, {1}")
```

File: cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/replacement/HostReplacementMultiDCInsufficientReplicasTest.java [hunk 4]
```diff
@@ -62,25 +61,10 @@ class HostReplacementMultiDCInsufficientReplicasTest extends HostReplacementTest
...
-        // Find the cause
-        while (cause != null && !StringUtils.contains(cause.getMessage(), "Failed to load"))
-        {
-            cause = cause.getCause();
-        }
-
-        assertThat(cause).isNotNull()
-                         .hasMessageFindingMatch("Failed to load (\\d+) ranges with EACH_QUORUM for " +
-                                                 "job ([a-zA-Z0-9-]+) in phase Environment Validation.");
+        sparkTestUtils.assertExpectedBulkWriteFailure(EACH_QUORUM.name(), dfWriter);
```
Comment (yifan-c) [inline cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/replacement/HostReplacementMultiDCInsufficientReplicasTest.java:67]: 👍

**Review by yifan-c — COMMENTED** [review]
looks good overall

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java (hunk 2 above)
Comment (frankgh) [reply cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java:498]: This is so we can customize the number of instances handled by Sidecar like we do here: https://github.com/apache/cassandra-analytics/pull/59/files#diff-a23a3d0a797ed790cbf779edbb5f6892cb31acff2d2f9bdc9c10a3347411b6d9R103

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java (hunk 1 above)
Comment (frankgh) [reply cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java]: The main reason to use String here is convenience. At `org.apache.cassandra.sidecar.testing.SharedClusterIntegrationTestBase` we use type `org.apache.cassandra.distributed.api.ConsistencyLevel`, but at the call sites we have other types available. In this case, we have type `com.datastax.driver.core.ConsistencyLevel` from the `org.apache.cassandra.analytics.TestConsistencyLevel`. Alternatively we can try to consolidate into a single CL type that we use across all of the tests. WDYT?

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java (hunk 2 above)
Comment (frankgh) [reply cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java:498]: The goal is to allow a Sidecar service to handle a subset of instances from the cluster. For that we don't expose the cluster itself, but rather only give the Sidecar module access to the instances it has to handle.

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java (hunk 1 above)
Comment (yifan-c) [reply cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java]: yep. "consolidate into a single CL type" sounds good and makes it easier to maintain in the future.

File: cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java (hunk 2 above)
Comment (yifan-c) [reply cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/sidecar/testing/SharedClusterIntegrationTestBase.java:498]: Now I see

**Pull request merged**